File: src/libs/ThirdParty.VBSuppliers/Properties/YouTravel/YouTravel.vb
Imports Intuitive
Imports Intuitive.Functions
Imports System.Text
Imports System.Xml
Imports Intuitive.Net.WebRequests
Imports ThirdParty.Abstractions
Imports ThirdParty.Abstractions.Models.Property.Booking
Imports ThirdParty.Abstractions.Models
Imports Intuitive.Helpers.Extensions
Imports ThirdParty.Abstractions.Constants
Imports ThirdParty.Abstractions.Lookups
Imports ThirdParty.Abstractions.Support
Imports System.Net.Http

Public Class YouTravel
    Implements IThirdParty

#Region "Constructor"

    Public Sub New(settings As IYouTravelSettings, support As ITPSupport, httpClient As HttpClient)
        _settings = Ensure.IsNotNull(settings, NameOf(settings))
        _support = Ensure.IsNotNull(support, NameOf(support))
        _httpClient = Ensure.IsNotNull(httpClient, NameOf(httpClient))
    End Sub

#End Region

#Region "Properties"

    Private Property _settings As IYouTravelSettings

    Private Property _support As ITPSupport

    Private Property _httpClient As HttpClient

    Public ReadOnly Property SupportsRemarks As Boolean Implements IThirdParty.SupportsRemarks
        Get
            Return True
        End Get
    End Property

    Private Function IThirdParty_SupportsLiveCancellation(searchDetails As IThirdPartyAttributeSearch, source As String) As Boolean Implements IThirdParty.SupportsLiveCancellation
        Return _settings.AllowCancellations(searchDetails, False)
    End Function

    Public ReadOnly Property SupportsBookingSearch() As Boolean Implements IThirdParty.SupportsBookingSearch
        Get
            Return False
        End Get
    End Property

    Private Function TakeSavingFromCommissionMargin(searchDetails As IThirdPartyAttributeSearch) As Boolean Implements IThirdParty.TakeSavingFromCommissionMargin
        Return False
    End Function

    Private Function OffsetCancellationDays(searchDetails As IThirdPartyAttributeSearch) As Integer Implements IThirdParty.OffsetCancellationDays
        Return _settings.OffsetCancellationDays(searchDetails, False)
    End Function

    Private Function IThirdParty_RequiresVCard(info As VirtualCardInfo) As Boolean Implements IThirdParty.RequiresVCard
        Return False
    End Function


#End Region

#Region "PreBook"

    Public Function PreBook(ByVal PropertyDetails As PropertyDetails) As Boolean Implements IThirdParty.PreBook
        Dim oResponse As New XmlDocument
        Dim bSuccess As Boolean = True
        Dim oWebRequest As New Request

        Try
            'Get Errata details
            Dim sb As New StringBuilder
            sb.AppendFormat("{0}{1}", _settings.PrebookURL(PropertyDetails), "?")
            sb.AppendFormat("&LangID={0}", _settings.LangID(PropertyDetails))
            sb.AppendFormat("&HID={0}", PropertyDetails.TPKey)
            sb.AppendFormat("&UserName={0}", _settings.Username(PropertyDetails))
            sb.AppendFormat("&Password={0}", _settings.Password(PropertyDetails))

            With oWebRequest
                .Source = ThirdParties.YOUTRAVEL
                .CreateLog = True
                .LogFileName = "Prebook"
                .EndPoint = sb.ToString
                .Method = eRequestMethod.GET
                .Send(_httpClient)
            End With

            oResponse = XMLFunctions.CleanXMLNamespaces(oWebRequest.ResponseXML)
            Dim oErrataNode As XmlNode = oResponse.SelectSingleNode("/HtSearchRq/Hotel/Erratas")
            Dim sErrata As String = oErrataNode.InnerText
            PropertyDetails.Errata.AddNew("Important Information", sErrata)
            'Get cancellation policies
            For Each oRoomDetails As RoomDetails In PropertyDetails.Rooms
                Dim oCancellationPolicyWebRequest As New Request
                Dim sbCancelPolicy As New StringBuilder
                Dim oCancelPolicyResponseXML As New XmlDocument
                sbCancelPolicy.AppendFormat("{0}{1}", _settings.CancellationPolicyURL(PropertyDetails), "?")
                sbCancelPolicy.AppendFormat("token={0}", oRoomDetails.ThirdPartyReference.Split("|"c)(2))
                With oCancellationPolicyWebRequest
                    .Source = ThirdParties.YOUTRAVEL
                    .CreateLog = True
                    .LogFileName = "Cancellation Cost"
                    .EndPoint = sbCancelPolicy.ToString
                    .Method = eRequestMethod.GET
                    .Send(_httpClient)
                End With

                oCancelPolicyResponseXML = XMLFunctions.CleanXMLNamespaces(oCancellationPolicyWebRequest.ResponseXML)
                For Each oCancellationNode As XmlNode In oCancelPolicyResponseXML.SelectNodes("/HtSearchRq/Policies/Policy")
                    Dim dFromDate As Date = oCancellationNode.SelectSingleNode("FromDate").InnerText.ToSafeDate
                    Dim dToDate As Date = New Date(2099, 1, 1)
                    Dim nCancellationCost As Decimal = SafeTypeExtensions.ToSafeDecimal(oCancellationNode.SelectSingleNode("Fees").InnerText)
                    PropertyDetails.Cancellations.AddNew(dFromDate, dToDate, nCancellationCost)
                Next
                PropertyDetails.Logs.AddNew(ThirdParties.YOUTRAVEL, "YouTravel CancellationPolicy Request " & oRoomDetails.PropertyRoomBookingID, oCancellationPolicyWebRequest.RequestLog)
                PropertyDetails.Logs.AddNew(ThirdParties.YOUTRAVEL, "YouTravel CancellationPolicy Response " & oRoomDetails.PropertyRoomBookingID, oCancellationPolicyWebRequest.ResponseLog)
            Next
            PropertyDetails.Cancellations.Solidify(SolidifyType.Sum)
        Catch ex As Exception
            bSuccess = False
            PropertyDetails.Warnings.AddNew("Prebook Exception", ex.ToString)
        Finally
            'save the xml for the front end
            PropertyDetails.Logs.AddNew(ThirdParties.YOUTRAVEL, "YouTravel Prebook Request", oWebRequest.RequestLog)
            PropertyDetails.Logs.AddNew(ThirdParties.YOUTRAVEL, "YouTravel Prebook Response", oWebRequest.ResponseLog)

        End Try

        PropertyDetails.LocalCost = PropertyDetails.Rooms.Sum(Function(r) r.LocalCost)

        Return bSuccess
    End Function

#End Region

#Region "Book"

    Public Function Book(ByVal PropertyDetails As PropertyDetails) As String Implements IThirdParty.Book

        Dim iSalesChannelID As Integer = PropertyDetails.SalesChannelID
        Dim iBrandID As Integer = PropertyDetails.BrandID
        Dim oResponse As New XmlDocument
        Dim sRequestURL As String = ""
        Dim sReference As String = ""

        Try

            'build url
            Dim sbURL As New StringBuilder
            sbURL.AppendFormat("{0}{1}", _settings.BookingURL(PropertyDetails), "?")
            sbURL.AppendFormat("&LangID={0}", _settings.LangID(PropertyDetails))
            sbURL.AppendFormat("&UserName={0}", _settings.Username(PropertyDetails))
            sbURL.AppendFormat("&Password={0}", _settings.Password(PropertyDetails))
            sbURL.AppendFormat("&session_ID={0}", PropertyDetails.Rooms(0).ThirdPartyReference.Split("|"c)(0))
            sbURL.AppendFormat("&Checkin_Date={0}", YouTravelSupport.FormatDate(PropertyDetails.ArrivalDate))
            sbURL.AppendFormat("&Nights={0}", PropertyDetails.Duration)
            sbURL.AppendFormat("&Rooms={0}", PropertyDetails.Rooms.Count)

            sbURL.AppendFormat("&HID={0}", PropertyDetails.TPKey)

            'adults and children
            Dim iRoomIndex As Integer = 0
            For Each oRoomDetails As RoomDetails In PropertyDetails.Rooms

                iRoomIndex += 1

                sbURL.AppendFormat("&ADLTS_{0}={1}", iRoomIndex, oRoomDetails.Adults)

                If oRoomDetails.Children + oRoomDetails.Infants > 0 Then

                    sbURL.AppendFormat("&CHILD_{0}={1}", iRoomIndex, oRoomDetails.Children + oRoomDetails.Infants)

                    For i As Integer = 1 To oRoomDetails.Children + oRoomDetails.Infants

                        If oRoomDetails.ChildAges.Count > i - 1 Then

                            sbURL.AppendFormat("&ChildAgeR{0}C{1}={2}", iRoomIndex, i, oRoomDetails.ChildAges(i - 1))

                        Else

                            sbURL.AppendFormat("&ChildAgeR{0}C{1}={2}", iRoomIndex, i, -1)

                        End If
                    Next

                Else

                    sbURL.AppendFormat("&CHILD_{0}=0", iRoomIndex)

                End If

                If iRoomIndex = 1 Then
                    sbURL.AppendFormat("&RID={0}", oRoomDetails.ThirdPartyReference.Split("|"c)(1))
                Else
                    sbURL.AppendFormat("&RID_{0}={1}", iRoomIndex, oRoomDetails.ThirdPartyReference.Split("|"c)(1))
                End If

                sbURL.AppendFormat("&Room{0}_Rate={1}", iRoomIndex, SafeMoney(oRoomDetails.GrossCost))

            Next

            sbURL.AppendFormat("&Customer_title={0}", PropertyDetails.Rooms(0).Passengers(0).Title)
            sbURL.AppendFormat("&Customer_firstname={0}", PropertyDetails.Rooms(0).Passengers(0).FirstName)
            sbURL.AppendFormat("&Customer_Lastname={0}", PropertyDetails.Rooms(0).Passengers(0).LastName)

            If PropertyDetails.BookingComments.Count > 0 Then
                'max 250 characters
                sbURL.AppendFormat("&Requests={0}", PropertyDetails.BookingComments.Item(0).Text.Substring _
                          (0, Min(PropertyDetails.BookingComments.Item(0).Text.Length, 250)))
            End If

            sRequestURL = sbURL.ToString

            Dim oWebRequest As New Request
            With oWebRequest
                .Source = ThirdParties.YOUTRAVEL
                .CreateLog = True
                .LogFileName = "Book"
                .EndPoint = sRequestURL
                .Method = eRequestMethod.GET
                .Send(_httpClient)
            End With

            oResponse = oWebRequest.ResponseXML

            'return booking reference if it exists
            Dim oNode As XmlNode = oResponse.SelectSingleNode("/HtSearchRq/Booking_ref")
            If oNode IsNot Nothing AndAlso Not oResponse.SelectSingleNode("/HtSearchRq/Booking_ref").InnerText = "" Then
                sReference = oResponse.SelectSingleNode("/HtSearchRq/Booking_ref").InnerText
            Else

                sReference = "failed"

            End If

        Catch ex As Exception

            PropertyDetails.Warnings.AddNew("Book Exception", ex.ToString)
            sReference = "failed"

        Finally

            'store the request and response xml on the property booking
            If sRequestURL <> "" Then
                PropertyDetails.Logs.AddNew(ThirdParties.YOUTRAVEL, "YouTravel Book Request", sRequestURL)
            End If

            If oResponse.InnerXml <> "" Then
                PropertyDetails.Logs.AddNew(ThirdParties.YOUTRAVEL, "YouTravel Book Response", oResponse)
            End If

        End Try

        Return sReference

    End Function

#End Region

#Region "Cancellation"

    Public Function CancelBooking(ByVal PropertyDetails As PropertyDetails) As ThirdPartyCancellationResponse Implements IThirdParty.CancelBooking

        Dim oThirdPartyCancellationResponse As New ThirdPartyCancellationResponse
        Dim sbURL As New StringBuilder
        Dim oResponse As New XmlDocument

        Try

            'build url
            sbURL.AppendFormat("{0}{1}", _settings.CancellationURL(PropertyDetails), "?")
            sbURL.AppendFormat("Booking_ref={0}", PropertyDetails.SourceReference)
            sbURL.AppendFormat("&UserName={0}", _settings.Username(PropertyDetails))
            sbURL.AppendFormat("&Password={0}", _settings.Password(PropertyDetails))


            'Send the request
            Dim oWebRequest As New Request
            With oWebRequest
                .Source = ThirdParties.YOUTRAVEL
                .CreateLog = True
                .LogFileName = "Cancel"
                .EndPoint = sbURL.ToString
                .Method = eRequestMethod.GET
                .Send(_httpClient)
            End With

            oResponse = oWebRequest.ResponseXML


            'check response
            If oResponse.InnerXml <> "" Then

                If oResponse.SelectSingleNode("/HtSearchRq/Success").InnerText = "True" Then

                    oThirdPartyCancellationResponse.TPCancellationReference = (oResponse.SelectSingleNode("/HtSearchRq/Booking_ref").InnerText)
                    oThirdPartyCancellationResponse.Success = True

                Else

                    oThirdPartyCancellationResponse.Success = False

                End If

            End If

        Catch ex As Exception

            PropertyDetails.Warnings.AddNew("Cancellation Exception", ex.ToString)
            oThirdPartyCancellationResponse.Success = False

        Finally

            'store the request and response xml on the property booking
            If sbURL.ToString <> "" Then
                PropertyDetails.Logs.AddNew(ThirdParties.YOUTRAVEL, "YouTravel Cancellation Request", sbURL.ToString)
            End If

            If oResponse.InnerXml <> "" Then
                PropertyDetails.Logs.AddNew(ThirdParties.YOUTRAVEL, "YouTravel Cancellation Response", oResponse)
            End If

        End Try

        Return oThirdPartyCancellationResponse

    End Function

    Public Function GetCancellationCost(ByVal PropertyDetails As PropertyDetails) As ThirdPartyCancellationFeeResult Implements IThirdParty.GetCancellationCost

        Dim oResult As New ThirdPartyCancellationFeeResult
        Dim iSalesChannelID As Integer = PropertyDetails.SalesChannelID
        Dim iBrandID As Integer = PropertyDetails.BrandID
        Dim sbURL As New StringBuilder
        Dim oResponse As New XmlDocument

        Try

            sbURL.AppendFormat("{0}{1}", _settings.CancellationFeeURL(PropertyDetails), "?")
            sbURL.AppendFormat("Booking_ref={0}", PropertyDetails.SourceReference)
            sbURL.AppendFormat("&UserName={0}", _settings.Username(PropertyDetails))
            sbURL.AppendFormat("&Password={0}", _settings.Password(PropertyDetails))


            'Send the request
            Dim oWebRequest As New Request
            With oWebRequest
                .Source = ThirdParties.YOUTRAVEL
                .CreateLog = True
                .LogFileName = "Cancellation Cost"
                .EndPoint = sbURL.ToString
                .Method = eRequestMethod.GET
                .Send(_httpClient)
            End With

            oResponse = oWebRequest.ResponseXML

            'return a fees result if found
            Dim oNode As XmlNode = oResponse.SelectSingleNode("/HtSearchRq/Success")
            If oNode IsNot Nothing AndAlso oResponse.SelectSingleNode("/HtSearchRq/Success").InnerText = "True" Then

                oResult.Success = True
                oResult.Amount = SafeDecimal(oResponse.SelectSingleNode("/HtSearchRq/Fees").InnerText)
                oResult.CurrencyCode = SafeString(oResponse.SelectSingleNode("/HtSearchRq/Currency").InnerText)

            Else

                oResult.Success = False

            End If

        Catch ex As Exception

            oResult.Success = False
            PropertyDetails.Warnings.AddNew("Cancellation Cost Exception", ex.ToString)

        Finally

            If sbURL.ToString <> "" Then
                PropertyDetails.Logs.AddNew(ThirdParties.YOUTRAVEL, "YouTravel Cancellation Cost Request", sbURL.ToString)
            End If

            If oResponse.InnerXml <> "" Then
                PropertyDetails.Logs.AddNew(ThirdParties.YOUTRAVEL, "YouTravel Cancellation Cost Response", oResponse)
            End If

        End Try

        Return oResult

    End Function

#End Region

#Region "Booking Search"

    Public Function BookingSearch(ByVal oBookingSearchDetails As BookingSearchDetails) As ThirdPartyBookingSearchResults Implements IThirdParty.BookingSearch
        Return New ThirdPartyBookingSearchResults
    End Function

    Public Function CreateReconciliationReference(ByVal sInputReference As String) As String Implements IThirdParty.CreateReconciliationReference
        Return ""
    End Function

#End Region

#Region "Booking Status Update"

    Public Function BookingStatusUpdate(ByVal oPropertyDetails As PropertyDetails) As ThirdPartyBookingStatusUpdateResult Implements IThirdParty.BookingStatusUpdate
        Return New ThirdPartyBookingStatusUpdateResult
    End Function

#End Region
    Public Sub EndSession(oPropertyDetails As PropertyDetails) Implements IThirdParty.EndSession

    End Sub
End Class

File: src/libs/ThirdParty.CSSuppliers/YouTravel/YouTravel.cs
namespace ThirdParty.CSSuppliers.YouTravel
{
    using System;
    using System.Linq;
    using System.Net.Http;
    using System.Text;
    using System.Threading.Tasks;
    using System.Xml;
    using Intuitive;
    using Intuitive.Helpers.Extensions;
    using Intuitive.Helpers.Serialization;
    using Intuitive.Net.WebRequests;
    using Microsoft.Extensions.Logging;
    using ThirdParty.Constants;
    using ThirdParty.Interfaces;
    using ThirdParty.Models;
    using ThirdParty.Models.Property.Booking;

    public class YouTravel : IThirdParty, ISingleSource
    {
        #region Constructor

        public YouTravel(IYouTravelSettings settings, HttpClient httpClient, ISerializer serializer, ILogger<YouTravel> logger)
        {
            _settings = Ensure.IsNotNull(settings, nameof(settings));
            _httpClient = Ensure.IsNotNull(httpClient, nameof(httpClient));
            _serializer = Ensure.IsNotNull(serializer, nameof(serializer));
            _logger = Ensure.IsNotNull(logger, nameof(logger));
        }

        #endregion

        #region Properties

        private readonly IYouTravelSettings _settings;

        private readonly HttpClient _httpClient;

        private readonly ISerializer _serializer;

        private readonly ILogger<YouTravel> _logger;

        public bool SupportsRemarks => true;

        public bool SupportsLiveCancellation(IThirdPartyAttributeSearch searchDetails, string source)
            => _settings.get_AllowCancellations(searchDetails, false);

        public bool SupportsBookingSearch => false;

        public string Source => ThirdParties.YOUTRAVEL;

        public int OffsetCancellationDays(IThirdPartyAttributeSearch searchDetails, string source)
            => _settings.get_OffsetCancellationDays(searchDetails, false);

        public bool RequiresVCard(VirtualCardInfo info, string source) => false;

        #endregion

        #region PreBook

        public async Task<bool> PreBookAsync(PropertyDetails propertyDetails)
        {
            var response = new XmlDocument();
            bool success = true;
            var webRequest = new Request();

            try
            {
                // Get Errata details
                var sb = new StringBuilder();
                sb.AppendFormat("{0}{1}", _settings.get_PrebookURL(propertyDetails), "?");
                sb.AppendFormat("&LangID={0}", _settings.get_LangID(propertyDetails));
                sb.AppendFormat("&HID={0}", propertyDetails.TPKey);
                sb.AppendFormat("&UserName={0}", _settings.get_Username(propertyDetails));
                sb.AppendFormat("&Password={0}", _settings.get_Password(propertyDetails));

                webRequest.Source = ThirdParties.YOUTRAVEL;
                webRequest.CreateLog = true;
                webRequest.LogFileName = "Prebook";
                webRequest.EndPoint = sb.ToString();
                webRequest.Method = eRequestMethod.GET;
                await webRequest.Send(_httpClient, _logger);

                response = _serializer.CleanXmlNamespaces(webRequest.ResponseXML);
                var errataNode = response.SelectSingleNode("/HtSearchRq/Hotel/Erratas");
                string errata = errataNode.InnerText;
                propertyDetails.Errata.AddNew("Important Information", errata);

                // Get cancellation policies
                foreach (var roomDetails in propertyDetails.Rooms)
                {
                    var sbCancelPolicy = new StringBuilder();
                    var cancelPolicyResponseXml = new XmlDocument();
                    sbCancelPolicy.AppendFormat("{0}{1}", _settings.get_CancellationPolicyURL(propertyDetails), "?");
                    sbCancelPolicy.AppendFormat("token={0}", roomDetails.ThirdPartyReference.Split('_')[2]);

                    var cancellationPolicyWebRequest = new Request
                    {
                        Source = ThirdParties.YOUTRAVEL,
                        CreateLog = true,
                        LogFileName = "Cancellation Cost",
                        EndPoint = sbCancelPolicy.ToString(),
                        Method = eRequestMethod.GET
                    };
                    await cancellationPolicyWebRequest.Send(_httpClient, _logger);

                    cancelPolicyResponseXml = _serializer.CleanXmlNamespaces(cancellationPolicyWebRequest.ResponseXML);
                    
                    foreach (XmlNode cancellationNode in cancelPolicyResponseXml.SelectNodes("/HtSearchRq/Policies/Policy"))
                    {
                        var fromDate = cancellationNode.SelectSingleNode("FromDate").InnerText.ToSafeDate();
                        var toDate = new DateTime(2099, 1, 1);
                        decimal cancellationCost = cancellationNode.SelectSingleNode("Fees").InnerText.ToSafeDecimal();
                        propertyDetails.Cancellations.AddNew(fromDate, toDate, cancellationCost);
                    }

                    propertyDetails.Logs.AddNew(ThirdParties.YOUTRAVEL, "YouTravel CancellationPolicy Request " + roomDetails.PropertyRoomBookingID, cancellationPolicyWebRequest.RequestLog);
                    propertyDetails.Logs.AddNew(ThirdParties.YOUTRAVEL, "YouTravel CancellationPolicy Response " + roomDetails.PropertyRoomBookingID, cancellationPolicyWebRequest.ResponseLog);
                }

                propertyDetails.Cancellations.Solidify(SolidifyType.Sum);
            }
            catch (Exception ex)
            {
                success = false;
                propertyDetails.Warnings.AddNew("Prebook Exception", ex.ToString());
            }
            finally
            {
                // save the xml for the front end
                propertyDetails.Logs.AddNew(ThirdParties.YOUTRAVEL, "YouTravel Prebook Request", webRequest.RequestLog);
                propertyDetails.Logs.AddNew(ThirdParties.YOUTRAVEL, "YouTravel Prebook Response", webRequest.ResponseLog);
            }

            propertyDetails.LocalCost = propertyDetails.Rooms.Sum(r => r.LocalCost);

            return success;
        }

        #endregion

        #region Book

        public async Task<string> BookAsync(PropertyDetails propertyDetails)
        {
            var response = new XmlDocument();
            string requestURL = "";
            string reference = "";

            try
            {
                // build url
                var sbURL = new StringBuilder();
                sbURL.AppendFormat("{0}{1}", _settings.get_BookingURL(propertyDetails), "?");
                sbURL.AppendFormat("&LangID={0}", _settings.get_LangID(propertyDetails));
                sbURL.AppendFormat("&UserName={0}", _settings.get_Username(propertyDetails));
                sbURL.AppendFormat("&Password={0}", _settings.get_Password(propertyDetails));
                sbURL.AppendFormat("&session_ID={0}", propertyDetails.Rooms[0].ThirdPartyReference.Split('|')[0]);
                sbURL.AppendFormat("&Checkin_Date={0}", YouTravelSupport.FormatDate(propertyDetails.ArrivalDate));
                sbURL.AppendFormat("&Nights={0}", propertyDetails.Duration);
                sbURL.AppendFormat("&Rooms={0}", propertyDetails.Rooms.Count);

                sbURL.AppendFormat("&HID={0}", propertyDetails.TPKey);

                // adults and children
                int roomIndex = 0;
                foreach (var roomDetails in propertyDetails.Rooms)
                {
                    roomIndex += 1;

                    sbURL.AppendFormat("&ADLTS_{0}={1}", roomIndex, roomDetails.Adults);

                    if (roomDetails.Children + roomDetails.Infants > 0)
                    {
                        sbURL.AppendFormat("&CHILD_{0}={1}", roomIndex, roomDetails.Children + roomDetails.Infants);

                        for (int i = 1; i <= roomDetails.Children + roomDetails.Infants; i++)
                        {
                            if (roomDetails.ChildAges.Count > i - 1)
                            {
                                sbURL.AppendFormat("&ChildAgeR{0}C{1}={2}", roomIndex, i, roomDetails.ChildAges[i - 1]);
                            }
                            else
                            {
                                sbURL.AppendFormat("&ChildAgeR{0}C{1}={2}", roomIndex, i, -1);
                            }
                        }
                    }
                    else
                    {
                        sbURL.AppendFormat("&CHILD_{0}=0", roomIndex);
                    }

                    if (roomIndex == 1)
                    {
                        sbURL.AppendFormat("&RID={0}", roomDetails.ThirdPartyReference.Split('|')[1]);
                    }
                    else
                    {
                        sbURL.AppendFormat("&RID_{0}={1}", roomIndex, roomDetails.ThirdPartyReference.Split('|')[1]);
                    }

                    sbURL.AppendFormat("&Room{0}_Rate={1}", roomIndex, roomDetails.GrossCost.ToSafeMoney());
                }

                sbURL.AppendFormat("&Customer_title={0}", propertyDetails.Rooms[0].Passengers[0].Title);
                sbURL.AppendFormat("&Customer_firstname={0}", propertyDetails.Rooms[0].Passengers[0].FirstName);
                sbURL.AppendFormat("&Customer_Lastname={0}", propertyDetails.Rooms[0].Passengers[0].LastName);

                if (propertyDetails.BookingComments.Count > 0)
                {
                    // max 250 characters
                    sbURL.AppendFormat("&Requests={0}", propertyDetails.BookingComments[0].Text.Substring(0, Math.Min(propertyDetails.BookingComments[0].Text.Length, 250)));
                }

                requestURL = sbURL.ToString();

                var webRequest = new Request
                {
                    Source = ThirdParties.YOUTRAVEL,
                    CreateLog = true,
                    LogFileName = "Book",
                    EndPoint = requestURL,
                    Method = eRequestMethod.GET
                };
                await webRequest.Send(_httpClient, _logger);

                response = webRequest.ResponseXML;

                // return booking reference if it exists
                var node = response.SelectSingleNode("/HtSearchRq/Booking_ref");
                if (node is not null && !string.IsNullOrEmpty(response.SelectSingleNode("/HtSearchRq/Booking_ref").InnerText))
                {
                    reference = response.SelectSingleNode("/HtSearchRq/Booking_ref").InnerText;
                }
                else
                {
                    reference = "failed";
                }
            }
            catch (Exception ex)
            {
                propertyDetails.Warnings.AddNew("Book Exception", ex.ToString());
                reference = "failed";
            }
            finally
            {
                // store the request and response xml on the property booking
                if (!string.IsNullOrEmpty(requestURL))
                {
                    propertyDetails.Logs.AddNew(ThirdParties.YOUTRAVEL, "YouTravel Book Request", requestURL);
                }

                if (!string.IsNullOrEmpty(response.InnerXml))
                {
                    propertyDetails.Logs.AddNew(ThirdParties.YOUTRAVEL, "YouTravel Book Response", response);
                }
            }

            return reference;
        }

        #endregion

        #region Cancellation

        public async Task<ThirdPartyCancellationResponse> CancelBookingAsync(PropertyDetails propertyDetails)
        {
            var thirdPartyCancellationResponse = new ThirdPartyCancellationResponse();
            var sbURL = new StringBuilder();
            var response = new XmlDocument();

            try
            {
                // build url
                sbURL.AppendFormat("{0}{1}", _settings.get_CancellationURL(propertyDetails), "?");
                sbURL.AppendFormat("Booking_ref={0}", propertyDetails.SourceReference);
                sbURL.AppendFormat("&UserName={0}", _settings.get_Username(propertyDetails));
                sbURL.AppendFormat("&Password={0}", _settings.get_Password(propertyDetails));

                // Send the request
                var webRequest = new Request
                {
                    Source = ThirdParties.YOUTRAVEL,
                    CreateLog = true,
                    LogFileName = "Cancel",
                    EndPoint = sbURL.ToString(),
                    Method = eRequestMethod.GET
                };
                await webRequest.Send(_httpClient, _logger);

                response = webRequest.ResponseXML;

                // check response
                if (!string.IsNullOrEmpty(response.InnerXml))
                {
                    if (response.SelectSingleNode("/HtSearchRq/Success").InnerText == "True")
                    {
                        thirdPartyCancellationResponse.TPCancellationReference = response.SelectSingleNode("/HtSearchRq/Booking_ref").InnerText;
                        thirdPartyCancellationResponse.Success = true;
                    }
                    else
                    {
                        thirdPartyCancellationResponse.Success = false;
                    }
                }
            }
            catch (Exception ex)
            {
                propertyDetails.Warnings.AddNew("Cancellation Exception", ex.ToString());
                thirdPartyCancellationResponse.Success = false;
            }
            finally
            {
                // store the request and response xml on the property booking
                if (!string.IsNullOrEmpty(sbURL.ToString()))
                {
                    propertyDetails.Logs.AddNew(ThirdParties.YOUTRAVEL, "YouTravel Cancellation Request", sbURL.ToString());
                }

                if (!string.IsNullOrEmpty(response.InnerXml))
                {
                    propertyDetails.Logs.AddNew(ThirdParties.YOUTRAVEL, "YouTravel Cancellation Response", response);
                }
            }

            return thirdPartyCancellationResponse;
        }

        public async Task<ThirdPartyCancellationFeeResult> GetCancellationCostAsync(PropertyDetails propertyDetails)
        {
            var result = new ThirdPartyCancellationFeeResult();
            var sbURL = new StringBuilder();
            var response = new XmlDocument();

            try
            {
                sbURL.AppendFormat("{0}{1}", _settings.get_CancellationFeeURL(propertyDetails), "?");
                sbURL.AppendFormat("Booking_ref={0}", propertyDetails.SourceReference);
                sbURL.AppendFormat("&UserName={0}", _settings.get_Username(propertyDetails));
                sbURL.AppendFormat("&Password={0}", _settings.get_Password(propertyDetails));

                // Send the request
                var webRequest = new Request
                {
                    Source = ThirdParties.YOUTRAVEL,
                    CreateLog = true,
                    LogFileName = "Cancellation Cost",
                    EndPoint = sbURL.ToString(),
                    Method = eRequestMethod.GET
                };
                await webRequest.Send(_httpClient, _logger);

                response = webRequest.ResponseXML;

                // return a fees result if found
                var node = response.SelectSingleNode("/HtSearchRq/Success");
                if (node is not null && response.SelectSingleNode("/HtSearchRq/Success").InnerText == "True")
                {
                    result.Success = true;
                    result.Amount = response.SelectSingleNode("/HtSearchRq/Fees").InnerText.ToSafeDecimal();
                    result.CurrencyCode = response.SelectSingleNode("/HtSearchRq/Currency").InnerText.ToSafeString();
                }
                else
                {
                    result.Success = false;
                }
            }
            catch (Exception ex)
            {
                result.Success = false;
                propertyDetails.Warnings.AddNew("Cancellation Cost Exception", ex.ToString());
            }
            finally
            {
                if (!string.IsNullOrEmpty(sbURL.ToString()))
                {
                    propertyDetails.Logs.AddNew(ThirdParties.YOUTRAVEL, "YouTravel Cancellation Cost Request", sbURL.ToString());
                }

                if (!string.IsNullOrEmpty(response.InnerXml))
                {
                    propertyDetails.Logs.AddNew(ThirdParties.YOUTRAVEL, "YouTravel Cancellation Cost Response", response);
                }
            }

            return result;
        }

        #endregion

        #region Booking Search

        public ThirdPartyBookingSearchResults BookingSearch(BookingSearchDetails bookingSearchDetails)
        {
            return new ThirdPartyBookingSearchResults();
        }

        public string CreateReconciliationReference(string sInputReference)
        {
            return "";
        }

        #endregion

        #region Booking Status Update

        public ThirdPartyBookingStatusUpdateResult BookingStatusUpdate(PropertyDetails propertyDetails)
        {
            return new ThirdPartyBookingStatusUpdateResult();
        }

        #endregion

        public void EndSession(PropertyDetails propertyDetails)
        {

        }
    }
}